service/cluster/cluster_test.go

```diff
@@ -0,0 +1,59 @@
+package cluster
+
+import (
+	"math/rand"
+	"reflect"
+	"testing"
+)
+
+func mockDiscoverPeers(peers []string, err error) func() ([]string, error) {
+	return func() ([]string, error) {
+		return peers, err
+	}
+}
+
+func TestGetPeers(t *testing.T) {
+	tests := []struct {
+		name              string
+		opts              Options
+		expectedPeers     []string
+		expectedError     error
+		discoverPeersMock func() ([]string, error)
+	}{
+		{
+			name:          "Test clustering disabled",
+			opts:          Options{EnableClustering: false},
+			expectedPeers: nil,
+		},
+		{
+			name:          "Test no max peers limit",
+			opts:          Options{EnableClustering: true, ClusterMaxJoinPeers: 0, DiscoverPeers: mockDiscoverPeers([]string{"A", "B"}, nil)},
+			expectedPeers: []string{"A", "B"},
+		},
+		{
+			name:          "Test max higher than number of peers",
+			opts:          Options{EnableClustering: true, ClusterMaxJoinPeers: 5, DiscoverPeers: mockDiscoverPeers([]string{"A", "B", "C"}, nil)},
+			expectedPeers: []string{"A", "B", "C"},
+		},
+		{
+			name:          "Test max peers limit with shuffling",
+			opts:          Options{EnableClustering: true, ClusterMaxJoinPeers: 2, DiscoverPeers: mockDiscoverPeers([]string{"A", "B", "C"}, nil)},
+			expectedPeers: []string{"A", "C"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := &Service{
+				opts:    test.opts,
+				randGen: rand.New(rand.NewSource(1)), // Seeded random generator to have consistent results in tests.
+			}
+
+			peers, _ := s.getPeers()
+
+			if !reflect.DeepEqual(peers, test.expectedPeers) {
+				t.Errorf("Expected peers %v, got %v", test.expectedPeers, peers)
+			}
```

Comment:
tip: you can use [require.ElementsMatch](https://pkg.go.dev/github.com/stretchr/testify/require#ElementsMatch) to assert that two slices are equal regardless of sort order, which can be useful for tests like these where the output order isn't guaranteed.